Flaky CI: "Type-check and test" job intermittently hangs to the 10-minute limit (orphan bun)

## Summary

The **"Type-check and test"** CI job intermittently **hangs to the 10-minute job limit and is canceled**, ending with `Terminate orphan process: pid (…) (bun)`. It is **flaky/infra, not a real test failure** — a plain re-run of the *same commit* passes in ~30s with no code change.

This is **not blocking any specific PR** (it was first dug into on #209), but it produces spurious red X's and wastes CI minutes, so it's worth tracking.

## Evidence

On PR #209, run [`26260691711`](https://github.com/buildinternet/releases-cli/actions/runs/26260691711):

| Attempt | Job | Result |
| --- | --- | --- |
| Original | [`77293247313`](https://github.com/buildinternet/releases-cli/actions/runs/26260691711/job/77293247313) | **canceled at 10m15s** ("exceeded the maximum execution time of 10m0s") |
| Re-run (no code change) | [`77296029045`](https://github.com/buildinternet/releases-cli/actions/runs/26260691711/job/77296029045) | **pass in 29s** |

In the failed run the log goes silent right after `agent-ergonomics.test.ts` test #2, then jumps 10 minutes to `##[error]The operation was canceled.` followed by `Terminate orphan process: pid (2443) (bun)`. The next test was `runCli([])` (bare `releases`).

It correlates with #207 ("auto-capture: mutation log + session traces") — that commit's own Test job also failed.

## What it is *not*

The bare-`releases` path is hermetic — nothing network/spawn/blocking runs for it:

- `checkForUpdate()` short-circuits on `.ts` source (`process.argv[1].endsWith(".ts")`), so no npm fetch under `runCli` (which runs `bun src/index.ts`).
- `maybeShowFirstRunNotice()` early-returns for the `internal-ci` client kind (CI sets `CI=true`).
- Telemetry `recordEvent()` is preempted by the default action's `process.exit(0)`.
- The completion notice is TTY-gated → `null` under piped stdio.
- trace/mutation-log aren't invoked without an API write; there are no global `process.on` exit handlers or timers in the startup path.

So the stall isn't a deterministic bug in the bare-`releases` code (or in the #209 changes).

## Mechanism: why a flake becomes a 10-minute hang

`runCli` (`tests/utils.ts`) uses `spawnSync` with `timeout: 30_000`. That timeout kills the *direct* child, but `spawnSync` keeps waiting on the stdout/stderr pipes until they close — so if the child (or anything it spawned, e.g. an inherited fd) holds a pipe open, the call blocks **past** 30s, all the way to the CI job limit.

```mermaid
flowchart TD
    A["test calls runCli([...])"] --> B["spawnSync('bun', ..., { timeout: 30000, stdio: pipe })"]
    B --> C{"child exits cleanly\nand closes pipes?"}
    C -- "yes (normal)" --> D["returns in ~0.2s"]
    C -- "no: a handle/pipe stays open" --> E["30s timeout fires,\nkills direct child"]
    E --> F{"pipe now closed?"}
    F -- "no (orphan still holds fd)" --> G["spawnSync keeps waiting on pipe\n→ blocks indefinitely"]
    G --> H["CI cancels job at 10m limit\n+ 'Terminate orphan process (bun)'"]
```

Locally this never reproduces (subprocesses exit fast), which is consistent with a network/runtime race that only surfaces on the CI runner.

## First action when you hit it

Re-run the job before assuming a code bug:

```bash
gh run rerun <run-id> --failed --repo buildinternet/releases-cli
```

## Suggested hardening (untried — can't verify without a repro)

1. **Make `runCli` subprocesses hermetic** so no background I/O can hold a pipe open:
   - `RELEASED_TELEMETRY_DISABLED=1`
   - `RELEASES_NO_COMPLETION_HINT=1`
2. **Make `spawnSync` fail fast instead of waiting on the pipe** — e.g. add `killSignal: "SIGKILL"` and/or drop the per-call timeout well below the CI job limit so a stuck child surfaces as a quick test failure (with a name) rather than a 10-minute job cancel.
3. **Investigate the #207 auto-capture path** (`src/lib/trace.ts` / `src/lib/mutation-log.ts`) for any fd/handle that can outlive a command, since the regression window lines up with it.

Because the hang isn't reproducible locally, any of the above should be treated as risk-reduction, not a confirmed fix — validate by watching CI stability over several runs.

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -74,7 +74,11 @@ jobs:
         run: npx tsc --noEmit
 
       - name: Test
-        run: bun test --isolate
+        # Single-process runner. `--isolate` (per-file VM isolation) was dropped
+        # in favor of de-leaking the mode.js mocks (see api-client.test.ts /
+        # search-lookup.test.ts) — its runner intermittently wedged a per-file
+        # isolate on the 2-vCPU runner, hanging the job to the 10m cap (#211).
+        run: bun test
 
   smoke-windows:
     name: Windows binary smoke test

--- a/package.json
+++ b/package.json
@@ -19,7 +19,7 @@
     "changeset": "changeset",
     "changeset:version": "changeset version && bun scripts/sync-version.ts && bun install",
     "changeset:publish": "changeset publish",
-    "test": "bun test --isolate",
+    "test": "bun test",
     "typecheck": "tsc --noEmit",
     "lint": "oxlint",
     "lint:fix": "oxlint --fix",

--- a/packages/lib/src/config.ts
+++ b/packages/lib/src/config.ts
@@ -4,10 +4,18 @@ import { homedir } from "os";
 import { legacyEnv } from "./legacy-env";
 
 let _dataDir: string | null = null;
+let _dataDirEnv: string | undefined;
 
 export function getDataDir(): string {
-  if (!_dataDir) {
-    _dataDir = legacyEnv("RELEASES_DATA_DIR", "RELEASED_DATA_DIR") || join(homedir(), ".releases");
+  // Cache the resolved dir, but invalidate when the env var changes. In prod
+  // the env is fixed at startup so this stays memoized (mkdir runs once); in
+  // tests, each file points RELEASED_DATA_DIR at its own temp dir, and the
+  // comparison re-resolves instead of returning a stale dir cached by another
+  // file. (This is what `bun test --isolate` used to paper over — see #211.)
+  const env = legacyEnv("RELEASES_DATA_DIR", "RELEASED_DATA_DIR");
+  if (_dataDir === null || env !== _dataDirEnv) {
+    _dataDirEnv = env;
+    _dataDir = env || join(homedir(), ".releases");
     mkdirSync(_dataDir, { recursive: true });
   }
   return _dataDir;

--- a/tests/unit/api-client.test.ts
+++ b/tests/unit/api-client.test.ts
@@ -1,15 +1,29 @@
-import { describe, it, expect, beforeEach, afterEach, mock } from "bun:test";
+import { describe, it, expect, beforeAll, afterAll, beforeEach, afterEach } from "bun:test";
 import { mkdtempSync, rmSync, readFileSync, existsSync } from "node:fs";
 import { tmpdir } from "node:os";
 import { join } from "node:path";
 import type { SourceWithOrg } from "../../src/api/types.js";
 
-// Mock mode.ts before importing the client — apiFetch calls getApiUrl/getApiKey.
-mock.module("../../src/lib/mode.js", () => ({
-  getApiUrl: () => "https://test.example.com",
-  getApiKey: () => "test-key",
-  isAdminMode: () => true,
-}));
+// Drive the real mode.ts via env rather than mocking the module. A top-level
+// mock.module("mode.js") is process-global — it leaks into every other test
+// file (e.g. mode-credential.test.ts), which was the sole reason the suite
+// needed `bun test --isolate`. apiFetch reads getApiUrl/getApiKey/isAdminMode
+// lazily, so setting env before the tests run reproduces the old stub
+// (getApiUrl → test URL, getApiKey → test-key, isAdminMode → true). Restored in
+// afterAll so the key never bleeds into resolveCredential tests.
+const prevEnv: { url?: string; key?: string } = {};
+beforeAll(() => {
+  prevEnv.url = process.env.RELEASES_API_URL;
+  prevEnv.key = process.env.RELEASES_API_KEY;
+  process.env.RELEASES_API_URL = "https://test.example.com";
+  process.env.RELEASES_API_KEY = "test-key";
+});
+afterAll(() => {
+  if (prevEnv.url === undefined) delete process.env.RELEASES_API_URL;
+  else process.env.RELEASES_API_URL = prevEnv.url;
+  if (prevEnv.key === undefined) delete process.env.RELEASES_API_KEY;
+  else process.env.RELEASES_API_KEY = prevEnv.key;
+});
 
 const client = await import("../../src/api/client.js");
 

--- a/tests/unit/search-lookup.test.ts
+++ b/tests/unit/search-lookup.test.ts
@@ -4,18 +4,30 @@
  * We capture console.log output and verify the lookup rail renders correctly
  * for each LookupStatus value.
  */
-import { describe, it, expect, mock, beforeEach, afterEach } from "bun:test";
+import { describe, it, expect, beforeAll, afterAll, beforeEach, afterEach } from "bun:test";
 import type { LookupResultPayload } from "../../src/api/types.js";
 
 // ---------------------------------------------------------------------------
-// Mock dependencies before importing the command module
+// Drive the real mode.ts via env instead of a process-global mock.module
+// (which leaked mode.js across files and forced `bun test --isolate`).
+// unifiedSearch → apiFetch reads getApiUrl/getApiKey lazily; the assertions
+// below only inspect the returned lookup payload, so the auth/admin values
+// don't matter. Restored in afterAll so the key never bleeds into other files.
 // ---------------------------------------------------------------------------
 
-mock.module("../../src/lib/mode.js", () => ({
-  getApiUrl: () => "https://test.example.com",
-  getApiKey: () => "test-key",
-  isAdminMode: () => false,
-}));
+const prevEnv: { url?: string; key?: string } = {};
+beforeAll(() => {
+  prevEnv.url = process.env.RELEASES_API_URL;
+  prevEnv.key = process.env.RELEASES_API_KEY;
+  process.env.RELEASES_API_URL = "https://test.example.com";
+  process.env.RELEASES_API_KEY = "test-key";
+});
+afterAll(() => {
+  if (prevEnv.url === undefined) delete process.env.RELEASES_API_URL;
+  else process.env.RELEASES_API_URL = prevEnv.url;
+  if (prevEnv.key === undefined) delete process.env.RELEASES_API_KEY;
+  else process.env.RELEASES_API_KEY = prevEnv.key;
+});
 
 // We'll inject a controlled response via fetch mock
 function mockSearch(response: unknown) {

